am-lib/src/main/java/uk/gov/hmcts/reform/amlib/AccessManagementService.java

```diff
@@ -68,27 +71,36 @@ public void revokeResourceAccess(ExplicitAccessMetadata explicitAccessMetadata)
      * @param userId       (accessorId)
      * @param resourceId   resource id
      * @param resourceJson json
-     * @return resourceJson or null
+     * @return filtered resourceJson based on READ permissions or null if no READ permissions on resource
      */
     public FilterResourceResponse filterResource(String userId, String resourceId, JsonNode resourceJson) {
-        ExplicitAccessRecord explicitAccess = jdbi.withExtension(AccessManagementRepository.class,
+        List<ExplicitAccessRecord> explicitAccess = jdbi.withExtension(AccessManagementRepository.class,
             dao -> dao.getExplicitAccess(userId, resourceId));
 
-        if (explicitAccess == null) {
+        if (explicitAccess.isEmpty()) {
             return null;
         }
 
-        if (READ.isGranted(explicitAccess.getPermissions())) {
-            Map<String, Set<Permission>> attributePermissions = new ConcurrentHashMap<>();
-            attributePermissions.put("/", Permissions.fromSumOf(explicitAccess.getPermissions()));
+        Map<String, Set<Permission>> attributePermissions = new ConcurrentHashMap<>();
+        ObjectNode filteredResource = JsonNodeFactory.instance.objectNode();
 
+        explicitAccess.forEach(explicitAccessRecord -> {
+            attributePermissions.put(explicitAccessRecord.getAttribute(),
+                Permissions.fromSumOf(explicitAccessRecord.getPermissions()));
+            if (READ.isGranted(explicitAccessRecord.getPermissions())) {
```

Comment:
Does it take visibility inheritance into consideration? Is for example possible to achieve following output for given permissions settings?

```
Permissions:

{
  "": [READ],
  "/name": [DELETE]
}

Output:

{
  resourceId: "#1",
  data: {
    "age": 18,
    "address": {
      "city": "London",
      "postcode": "SE1"
    }
  },
  permissions: {
    "": [READ],
    "/name": [DELETE]
  }
}
```

Reply:
Since filtering is getting more complex having whole logic in the service does not make much sense. I suggest extracting JSON filtering logic to separate class.